src/Microsoft.Health.SqlServer/Features/Client/SqlConnectionWrapper.cs

```diff
@@ -17,55 +16,74 @@ public class SqlConnectionWrapper : IDisposable
         private readonly bool _enlistInTransactionIfPresent;
         private readonly SqlTransactionHandler _sqlTransactionHandler;
         private readonly SqlCommandWrapperFactory _sqlCommandWrapperFactory;
+        private readonly ISqlConnectionFactory _sqlConnectionFactory;
+
+        private SqlConnection _sqlConnection;
+        private SqlTransaction _sqlTransaction;
 
         public SqlConnectionWrapper(
-            SqlServerDataStoreConfiguration configuration,
             SqlTransactionHandler sqlTransactionHandler,
             SqlCommandWrapperFactory sqlCommandWrapperFactory,
+            ISqlConnectionFactory sqlConnectionFactory,
             bool enlistInTransactionIfPresent)
         {
-            EnsureArg.IsNotNull(configuration, nameof(configuration));
             EnsureArg.IsNotNull(sqlTransactionHandler, nameof(sqlTransactionHandler));
             EnsureArg.IsNotNull(sqlCommandWrapperFactory, nameof(sqlCommandWrapperFactory));
+            EnsureArg.IsNotNull(sqlConnectionFactory, nameof(sqlConnectionFactory));
 
             _sqlTransactionHandler = sqlTransactionHandler;
             _enlistInTransactionIfPresent = enlistInTransactionIfPresent;
             _sqlCommandWrapperFactory = sqlCommandWrapperFactory;
+            _sqlConnectionFactory = sqlConnectionFactory;
+
+            InitializeAsync();
```

Comment:
You should not call async method from consturctor.

Reply:
The behavior is underministic because you are not waiting for the operation to finish before returning it so depending on how thread pool manages the tasks, it might not be completed before the caller access SqlConnectionproperty.

What I was suggesting is to make SqlConnectionWrapperFactory.ObtainSqlConnectionWrapper async call and within there, you can create an instance and also call InitializeAsync. I would make the constructor of SqlConnectionWrapper and InitializeAsync internal instead of public so that the life cycle management is completely done through SqlConnectionWrapperFactory which will always initializes it properly.

Reply:
Ahaan I misunderstood it earlier. This makes sense. I will make the changes and publish the updated iteration shortly.